File: backend/src/main/java/com/blissfuljuan/aiprojecteval/projectproposal/service/ProjectProposalService.java
package com.blissfuljuan.aiprojecteval.projectproposal.service;

import com.blissfuljuan.aiprojecteval.document.dto.DocumentResponse;
import com.blissfuljuan.aiprojecteval.projectproposal.dto.AdviserDecisionRequest;
import com.blissfuljuan.aiprojecteval.projectproposal.dto.ProjectProposalDocumentLinkRequest;
import com.blissfuljuan.aiprojecteval.projectproposal.dto.ProjectProposalCreateRequest;
import com.blissfuljuan.aiprojecteval.projectproposal.dto.ProjectProposalResponse;
import com.blissfuljuan.aiprojecteval.projectproposal.dto.ProjectProposalUpdateRequest;
import com.blissfuljuan.aiprojecteval.projectproposal.dto.ProposalDecisionRequest;
import java.util.List;

public interface ProjectProposalService {

	ProjectProposalResponse createProposal(String currentUserEmail, ProjectProposalCreateRequest request);

	List<ProjectProposalResponse> getMyProposals(String currentUserEmail);

	List<ProjectProposalResponse> getAllProposals(String currentUserEmail);

	ProjectProposalResponse getProposalById(String currentUserEmail, Long id);

	ProjectProposalResponse updateProposal(String currentUserEmail, Long id, ProjectProposalUpdateRequest request);

	void deleteProposal(String currentUserEmail, Long id);

	ProjectProposalResponse adviserDecision(String currentUserEmail, Long id, AdviserDecisionRequest request);

	ProjectProposalResponse instructorDecision(String currentUserEmail, Long id, ProposalDecisionRequest request);

	List<DocumentResponse> getProposalDocuments(String currentUserEmail, Long id);

	DocumentResponse submitProposalDocumentLink(
			String currentUserEmail,
			Long id,
			ProjectProposalDocumentLinkRequest request);

	void validateCanViewProposalDocuments(String currentUserEmail, Long id);

	void validateCanManageProposalDocuments(String currentUserEmail, Long id);
}

File: web/src/modules/project-proposal/services/projectProposal.service.ts
import { isAxiosError } from "axios";
import { api } from "@/common/lib/api";
import type { GenericDocument } from "@/modules/document/types";
import type { ApiResponse } from "@/modules/identity/types";
import type {
  AdviserDecisionRequest,
  InstructorDecisionRequest,
  ProjectProposal,
  ProjectProposalDocumentLinkRequest,
  ProposalCreateRequest,
  ProposalUpdateRequest,
} from "@/modules/project-proposal/types";

async function createProposal(request: ProposalCreateRequest): Promise<ProjectProposal> {
  const response = await api.post<ApiResponse<ProjectProposal>>("/api/project-proposals", request);
  return response.data.data;
}

async function getMyProposals(): Promise<ProjectProposal[]> {
  const response = await api.get<ApiResponse<ProjectProposal[]>>("/api/project-proposals/my");
  return response.data.data;
}

async function getAllProposals(): Promise<ProjectProposal[]> {
  const response = await api.get<ApiResponse<ProjectProposal[]>>("/api/project-proposals");
  return response.data.data;
}

async function getProposalById(id: number): Promise<ProjectProposal> {
  const response = await api.get<ApiResponse<ProjectProposal>>(`/api/project-proposals/${id}`);
  return response.data.data;
}

async function updateProposal(id: number, request: ProposalUpdateRequest): Promise<ProjectProposal> {
  const response = await api.put<ApiResponse<ProjectProposal>>(`/api/project-proposals/${id}`, request);
  return response.data.data;
}

async function deleteProposal(id: number): Promise<void> {
  await api.delete(`/api/project-proposals/${id}`);
}

async function adviserDecision(id: number, request: AdviserDecisionRequest): Promise<ProjectProposal> {
  const response = await api.patch<ApiResponse<ProjectProposal>>(
    `/api/project-proposals/${id}/adviser-decision`,
    request,
  );
  return response.data.data;
}

async function instructorDecision(id: number, request: InstructorDecisionRequest): Promise<ProjectProposal> {
  const response = await api.patch<ApiResponse<ProjectProposal>>(
    `/api/project-proposals/${id}/instructor-decision`,
    request,
  );
  return response.data.data;
}

async function getProposalDocuments(id: number): Promise<GenericDocument[]> {
  const response = await api.get<ApiResponse<GenericDocument[]>>(`/api/project-proposals/${id}/documents`);
  return response.data.data;
}

async function submitProposalDocumentLink(
  id: number,
  request: ProjectProposalDocumentLinkRequest,
): Promise<GenericDocument> {
  const response = await api.post<ApiResponse<GenericDocument>>(`/api/project-proposals/${id}/documents/link`, request);
  return response.data.data;
}

function getErrorMessage(error: unknown): string {
  if (isAxiosError<ApiResponse<unknown>>(error)) {
    const data = error.response?.data;
    return data?.errors?.[0] ?? data?.message ?? "Request failed. Please try again.";
  }
  if (error instanceof Error) return error.message;
  return "Request failed. Please try again.";
}

export const projectProposalService = {
  createProposal,
  getMyProposals,
  getAllProposals,
  getProposalById,
  updateProposal,
  deleteProposal,
  adviserDecision,
  instructorDecision,
  getProposalDocuments,
  submitProposalDocumentLink,
  getErrorMessage,
};
